### ISSUE TITLE
why inference befor optimizer.zero_grad()?

### ISSUE THREAD

**Issue by guyii54** [thuml/Transfer-Learning-Library#12]
for example, in ./exapmle/dann.py, line 164 and 165.
```
optimizer.zero_grad()
loss.backward()
optimizer.step()
```
The inference and calculation of loss is before optimizer.zero_grad(), then optimizer.zero_grad() sets the grads to zero, in this situation, whay the loss can backward()?

**Comment by microhhh** [comment]
This is because by default, gradients are accumulated in buffers (i.e, not overwritten) whenever .backward() is called.

You can learn the Deep Learning and PyTorch carefully from this book: https://pytorch.org/deep-learning-with-pytorch-thank-you

**Comment by microhhh** [comment]
When backpropagating, we calculate the gradient of parameters.

**Comment by guyii54** [comment]
I've got it, thanks for you reply
